src/main/java/net/snowflake/ingest/connection/OAuthClient.java

```diff
@@ -4,11 +4,122 @@
 
-/** Interface to perform token refresh request from {@link OAuthManager} */
-public interface OAuthClient {
-  AtomicReference<OAuthCredential> getoAuthCredentialRef();
+/*
+ * Implementation of OAuth Client, used for refreshing an OAuth access token.
+ */
+public class OAuthClient {
+
+  static final Logger LOGGER = LoggerFactory.getLogger(OAuthClient.class);
+  private static final String TOKEN_REQUEST_ENDPOINT = "/oauth/token-request";
+
+  // Content type header to specify the encoding
+  private static final String OAUTH_CONTENT_TYPE_HEADER = "application/x-www-form-urlencoded";
+  private static final String GRANT_TYPE_PARAM = "grant_type";
+  private static final String ACCESS_TOKEN = "access_token";
+  private static final String REFRESH_TOKEN = "refresh_token";
+  private static final String EXPIRES_IN = "expires_in";
+
+  // OAuth credential
+  private final AtomicReference<OAuthCredential> oAuthCredential;
+
+  // Http client for submitting token refresh request
+  private final CloseableHttpClient httpClient;
+
+  /**
+   * Creates an AuthClient for Snowflake OAuth given account, credential and base uri
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param oAuthCredential - the OAuth credential we're using to connect
+   * @param baseURIBuilder - the uri builder with common scheme, host and port
+   */
+  OAuthClient(String accountName, OAuthCredential oAuthCredential, URIBuilder baseURIBuilder) {
+    this.oAuthCredential = new AtomicReference<>(oAuthCredential);
+
+    // build token request uri
+    baseURIBuilder.setPath(TOKEN_REQUEST_ENDPOINT);
+    this.httpClient = HttpUtil.getHttpClient(accountName);
+  }
+
+  /** Get access token */
+  public AtomicReference<OAuthCredential> getOAuthCredentialRef() {
+    return oAuthCredential;
+  }
+
+  /** Refresh access token using a valid refresh token */
+  public void refreshToken() {
+    String respBodyString = null;
+    try (CloseableHttpResponse httpResponse = httpClient.execute(makeRefreshTokenRequest())) {
+      respBodyString = EntityUtils.toString(httpResponse.getEntity());
+
+      if (httpResponse.getStatusLine().getStatusCode() == HttpStatusCodes.STATUS_CODE_OK) {
+        JsonObject respBody = JsonParser.parseString(respBodyString).getAsJsonObject();
+
+        if (respBody.has(ACCESS_TOKEN) && respBody.has(EXPIRES_IN)) {
+          // Trim surrounding quotation marks
+          String newAccessToken = respBody.get(ACCESS_TOKEN).toString().replaceAll("^\"|\"$", "");
+
+          oAuthCredential.get().setAccessToken(newAccessToken);
+          oAuthCredential.get().setExpiresIn(respBody.get(EXPIRES_IN).getAsInt());
+          return;
+        }
+      }
+    } catch (IOException e) {
+      throw new SFException(ErrorCode.OAUTH_REFRESH_TOKEN_ERROR, e.getMessage());
+    }
+    throw new SFException(
+            ErrorCode.OAUTH_REFRESH_TOKEN_ERROR,
+            "Refresh access token fail with response: " + respBodyString);
```

Comment:
ok, I think now this can be remove to the else branch because you're not catching all exceptions, basically
try 
if (status == ok) else (throw SFException)
catch IOException
throw SFException

Reply:
Got it, fixed. Thank you!